ssl/statem/extensions.c

```diff
@@ -1684,7 +1684,7 @@ static int init_post_handshake_auth(SSL *s, ossl_unused unsigned int context)
  */
 static int final_psk(SSL *s, unsigned int context, int sent)
 {
-    if (s->server && sent && s->clienthello != NULL
+    if (SSL_IS_TLS13(s) && s->server && sent && s->clienthello != NULL
```

Comment:
While this is "harmless", I don't think it should be necessary.
"pre_shared_key" is a TLS 1.3-only extension, and we only enter this block (and hit the `SSLfatal()`) if the client sent the extension (`sent` is true).
If the client sends "pre_shared_key" that means the client knows about what RFC 8446 says, and the requirement from https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.9 that a client MUST provide "psk_key_exchange_modes" if it offers "pre_shared_key" applies.

A client has no reason to send "pre_shared_key" in a renegotiation handshake (since the client knows that TLS 1.3 will not be negotiated and thus the PSK is useless), but if the client sends "pre_shared_key" the specification for "pre_shared_key" (RFC 8446) applies, and therefore this check as well.

Reply:
(Ben asked me to comment here as well. I agree that this is odd. Slightly longer thoughts at https://github.com/openssl/openssl/issues/15375#issuecomment-845424484.)